File: packages/sdk/arc-go/invariants/json.go
package invariants

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"strconv"
	"strings"
	"unicode/utf16"
)

func CanonicalizeJSONString(input string) (string, error) {
	value, err := ParseJSONText(input)
	if err != nil {
		return "", err
	}
	return CanonicalizeJSON(value)
}

func CanonicalizeJSON(value any) (string, error) {
	switch typed := value.(type) {
	case nil:
		return "null", nil
	case bool:
		if typed {
			return "true", nil
		}
		return "false", nil
	case string:
		return marshalJSONString(typed), nil
	case json.Number:
		return canonicalizeJSONNumber(typed)
	case float64:
		return canonicalizeFloat(typed)
	case float32:
		return canonicalizeFloat(float64(typed))
	case int:
		return strconv.Itoa(typed), nil
	case int8:
		return strconv.FormatInt(int64(typed), 10), nil
	case int16:
		return strconv.FormatInt(int64(typed), 10), nil
	case int32:
		return strconv.FormatInt(int64(typed), 10), nil
	case int64:
		return strconv.FormatInt(typed, 10), nil
	case uint:
		return strconv.FormatUint(uint64(typed), 10), nil
	case uint8:
		return strconv.FormatUint(uint64(typed), 10), nil
	case uint16:
		return strconv.FormatUint(uint64(typed), 10), nil
	case uint32:
		return strconv.FormatUint(uint64(typed), 10), nil
	case uint64:
		return strconv.FormatUint(typed, 10), nil
	case []any:
		parts := make([]string, 0, len(typed))
		for _, item := range typed {
			rendered, err := CanonicalizeJSON(item)
			if err != nil {
				return "", err
			}
			parts = append(parts, rendered)
		}
		return "[" + strings.Join(parts, ",") + "]", nil
	case map[string]any:
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sortUTF16(keys)
		parts := make([]string, 0, len(keys))
		for _, key := range keys {
			rendered, err := CanonicalizeJSON(typed[key])
			if err != nil {
				return "", err
			}
			parts = append(parts, fmt.Sprintf("%s:%s", marshalJSONString(key), rendered))
		}
		return "{" + strings.Join(parts, ",") + "}", nil
	default:
		return "", newInvariantError("canonical_json", fmt.Sprintf(
			"canonical JSON does not support values of type %T",
			value,
		))
	}
}

func canonicalizeJSONNumber(number json.Number) (string, error) {
	raw := number.String()
	if !strings.ContainsAny(raw, ".eE") {
		if raw == "-0" {
			return "0", nil
		}
		return raw, nil
	}
	parsed, err := strconv.ParseFloat(raw, 64)
	if err != nil {
		return "", newInvariantError("canonical_json", "canonical JSON does not support invalid numbers")
	}
	return canonicalizeFloat(parsed)
}

func canonicalizeFloat(value float64) (string, error) {
	if math.IsNaN(value) || math.IsInf(value, 0) {
		return "", newInvariantError("canonical_json", "canonical JSON does not support non-finite numbers")
	}
	if value == 0 {
		return "0", nil
	}
	if math.Trunc(value) == value && math.Abs(value) < 1e21 {
		return strconv.FormatFloat(value, 'f', 0, 64), nil
	}
	rendered := strings.ToLower(strconv.FormatFloat(value, 'g', -1, 64))
	return normalizeExponent(rendered), nil
}

func normalizeExponent(value string) string {
	index := strings.IndexByte(value, 'e')
	if index == -1 {
		return value
	}
	mantissa := value[:index]
	exponent := value[index+1:]
	if len(exponent) < 2 {
		return value
	}
	sign := exponent[:1]
	digits := strings.TrimLeft(exponent[1:], "0")
	if digits == "" {
		digits = "0"
	}
	return mantissa + "e" + sign + digits
}

func marshalJSONString(input string) string {
	var buffer bytes.Buffer
	encoder := json.NewEncoder(&buffer)
	encoder.SetEscapeHTML(false)
	_ = encoder.Encode(input)
	return strings.TrimSuffix(buffer.String(), "\n")
}

func sortUTF16(values []string) {
	for left := 0; left < len(values); left += 1 {
		for right := left + 1; right < len(values); right += 1 {
			if compareUTF16(values[right], values[left]) < 0 {
				values[left], values[right] = values[right], values[left]
			}
		}
	}
}

func compareUTF16(left string, right string) int {
	leftUnits := utf16.Encode([]rune(left))
	rightUnits := utf16.Encode([]rune(right))
	for index := 0; index < len(leftUnits) && index < len(rightUnits); index += 1 {
		if leftUnits[index] < rightUnits[index] {
			return -1
		}
		if leftUnits[index] > rightUnits[index] {
			return 1
		}
	}
	if len(leftUnits) < len(rightUnits) {
		return -1
	}
	if len(leftUnits) > len(rightUnits) {
		return 1
	}
	return 0
}

File: sdks/python/chio-py/src/chio/invariants/json.py
from __future__ import annotations

import json
import math
import re
from decimal import Decimal
from typing import Any

from ..errors import ChioInvariantError, parse_json_text

_EXPONENT_RE = re.compile(r"e([+-])0+(\d+)$")


def _canonical_json_string(value: str) -> str:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def _canonicalize_float(value: float) -> str:
    if not math.isfinite(value):
        raise ChioInvariantError(
            "canonical_json",
            "canonical JSON does not support non-finite numbers",
        )
    if value == 0.0:
        return "0"
    rendered = repr(value).lower()
    if 1e-6 <= abs(value) < 1e21:
        decimal = format(Decimal(rendered), "f")
        if "." in decimal:
            decimal = decimal.rstrip("0").rstrip(".")
        return decimal
    return _EXPONENT_RE.sub(lambda match: f"e{match.group(1)}{match.group(2)}", rendered)


def canonicalize_json(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        return _canonicalize_float(value)
    if isinstance(value, str):
        return _canonical_json_string(value)
    if isinstance(value, list):
        return "[" + ",".join(canonicalize_json(item) for item in value) + "]"
    if isinstance(value, dict):
        if not all(isinstance(key, str) for key in value):
            raise ChioInvariantError("canonical_json", "canonical JSON object keys must be strings")
        items = sorted(value.items(), key=lambda item: item[0].encode("utf-16-be"))
        return "{" + ",".join(
            f"{_canonical_json_string(key)}:{canonicalize_json(entry_value)}"
            for key, entry_value in items
        ) + "}"
    raise ChioInvariantError(
        "canonical_json",
        f"canonical JSON does not support values of type {type(value).__name__}",
    )


def canonicalize_json_string(input_text: str) -> str:
    return canonicalize_json(parse_json_text(input_text))
